### PR TITLE
feat: support resolve default value when environment not set

### PR DESCRIPTION
**Pull request by kevinw66**
Thanks for review my first pull request for apisix.
And this is my first time use lua. Give me some advices if possible, thanks.

This is about read default value when ENV not set, I think this may helpful when using k8s in different environment and did not want to set so many ENV variable

result:
etcd.host = http://${{ETCD_HOST}}:2379

image

etcd.host = http://${{ETCD_HOST:123ss}}:2379
image

etcd.host = http://${{ETCD_HOST:123ss:123aa}}:2379
image

### CODE REVIEW

File: apisix/cli/file.lua [hunk 1]
```diff
@@ -65,8 +65,15 @@ local function resolve_conf_var(conf)
             local var_used = false
             -- we use '${{var}}' because '$var' and '${var}' are taken
             -- by Nginx
-            local new_val = val:gsub("%$%{%{%s*([%w_]+)%s*%}%}", function(var)
-                local v = getenv(var)
+            local new_val = val:gsub("%$%{%{%s*([%w_]+[%:%w+]*)%s*%}%}", function(var)
```
Comment (spacewander) [inline apisix/cli/file.lua:68]: Could you add a test in https://github.com/apache/apisix/blob/b8daf65b3c0690138feccb2ce399fc58349ac7ee/t/cli/test_main.sh#L311?
Also need to update the doc in https://github.com/apache/apisix/blob/11433d740c5e19b757c61f59842d6de64e27fba8/conf/config.yaml#L24

Comment (kevinw66) [reply apisix/cli/file.lua:68]: OK, I'll try

Comment (kevinw66) [reply apisix/cli/file.lua:68]: Done

File: t/cli/test_main.sh [hunk 2]
```diff
@@ -310,6 +310,35 @@ fi
...
+echo '
+etcd:
+    host:
+        - "http://${{ETCD_HOST:192.168.1.1}}:${{ETCD_PORT}}"
+' > conf/config.yaml
+
+ETCD_PORT=2379 make init
+
+if ! grep "env ETCD_HOST=192.168.1.1;" conf/nginx.conf > /dev/null; then
+    echo "failed: support default value when environment not set"
```
Comment (tzssangglass) [inline t/cli/test_main.sh:323]:
```suggestion
    echo "failed: should use default value when environment not set"
```

File: t/cli/test_main.sh [hunk 3]
```diff
@@ -310,6 +310,35 @@ fi
...
+echo '
+etcd:
+    host:
+        - "http://${{ETCD_HOST:192.168.1.1}}:${{ETCD_PORT}}"
+' > conf/config.yaml
+
+ETCD_HOST=127.0.0.1 ETCD_PORT=2379 make init
+
+if ! grep "env ETCD_HOST=127.0.0.1;" conf/nginx.conf > /dev/null; then
+    echo "failed: support default value when environment not set"
```
Comment (tzssangglass) [inline t/cli/test_main.sh:336]:
```suggestion
    echo "failed: should use  environment var of ETCD_HOST"
```

File: t/cli/test_main.sh [hunk 4]
```diff
@@ -310,6 +310,35 @@ fi
 
 echo "pass: support environment variables in local_conf"
 
+# support default value when environment not set
+echo '
+etcd:
+    host:
+        - "http://${{ETCD_HOST:192.168.1.1}}:${{ETCD_PORT}}"
```
Comment (tzssangglass) [inline t/cli/test_main.sh:317]: Can we use an other name? I'm worried that ETCD_HOST may have already been set with variables.

File: t/cli/test_main.sh [hunk 5]
```diff
@@ -310,6 +310,35 @@ fi
...
+' > conf/config.yaml
+
+ETCD_PORT=2379 make init
+
+if ! grep "env ETCD_HOST=192.168.1.1;" conf/nginx.conf > /dev/null; then
```
Comment (tzssangglass) [inline t/cli/test_main.sh:322]: I think it would be more appropriate to check if etcd is connected properly from the APISIX error.log.

**Review by tzssangglass — APPROVED** [review]

File: t/cli/test_main.sh (hunk 4 above)
Comment (kevinw66) [reply t/cli/test_main.sh:317]: But ETCD_HOST is used in previous tests, so I reuse this variable

Comment (tzssangglass) [reply t/cli/test_main.sh:317]: OK, let follow this.

File: t/cli/test_main.sh (hunk 5 above)
Comment (kevinw66) [reply t/cli/test_main.sh:322]: When etcd cannot connect correctly, I found that error will print in console, not error.log

**Comment by kevinw66** [comment]
The github ci test result in `Error: [error] open() "/home/runner/work/apisix/apisix/logs/nginx.pid" failed (2: No such file or directory)` is this my code problem ? I saw it failed in other testcases.

File: conf/config.yaml [hunk 6]
```diff
@@ -30,6 +30,16 @@
 # And then run `export ETCD_HOST=$your_host` before `make init`.
 #
 # If the configured environment variable can't be found, an error will be thrown.
+#
+# Also, If you want to use default value when the environment variable not set,
+# Use `${{VAR::default_value}}` instead. For instance:
```
Comment (spacewander) [inline conf/config.yaml:35]: I am curious why `::` is chosen as the separator. It is different from the common way. Is there a special reason behind it?

File: apisix/cli/file.lua [hunk 7]
```diff
@@ -65,8 +65,15 @@ local function resolve_conf_var(conf)
             local var_used = false
             -- we use '${{var}}' because '$var' and '${var}' are taken
             -- by Nginx
-            local new_val = val:gsub("%$%{%{%s*([%w_]+)%s*%}%}", function(var)
-                local v = getenv(var)
+            local new_val = val:gsub("%$%{%{%s*([%w_]+[%:%:[%w%-%.%:%?%/=]+]?)%s*%}%}", function(var)
```
Comment (spacewander) [inline apisix/cli/file.lua:68]: Why do we need a regex to check the default variable?

File: conf/config.yaml (hunk 6 above)
Comment (kevinw66) [reply conf/config.yaml:35]: I want to use `:` in first time, But sometimes default value may contains it, like etcd.host=${{ETCD_HOST::demo.domain.com:1234}}

File: apisix/cli/file.lua (hunk 7 above)
Comment (kevinw66) [reply apisix/cli/file.lua:68]: If we don't check, It will fail to match when config contains default value.

**Comment by kevinw66** [comment]
BTW, Can you help me to check the ut? I'm confusing about the error log, and it cannot run in my local machine right now.

File: conf/config.yaml (hunk 6 above)
Comment (spacewander) [reply conf/config.yaml:35]: What about using `:=`? In bash, it means "Assign  Default Values".

File: apisix/cli/file.lua (hunk 7 above)
Comment (spacewander) [reply apisix/cli/file.lua:68]: Sure. But we don't need to be too strict with `[%w%-%.%:%?%/=]+`? What if people want to use space in the default value?

**Comment by spacewander** [comment]
Solving with https://github.com/apache/apisix/pull/5699

File: t/cli/test_main.sh [hunk 8]
```diff
@@ -310,6 +310,45 @@ fi
...
+' > conf/config.yaml
+
+TEST_ENV=127.0.0.1 make init
+
+if ! grep "env TEST_ENV=127.0.0.1;" conf/nginx.conf > /dev/null; then
+    echo "failed: should use default value when environment not set"
+    exit 1
+fi
+
+echo "pass: should use default value when environment not set"
```
Comment (tzssangglass) [inline t/cli/test_main.sh:350]: we can use `support default value when environment not set` here. Earlier I suggested taht change `echo "failed: support default value when environment not set"` to `echo "failed: should use default value when environment not set"` above is to output exactly that failed test case when it fails. If the error output is the same for all three tests above, then when an error occurs we need to take the time to see which one is actually wrong.

I think you can tweak the error output of the test case a little.

Comment (kevinw66) [reply t/cli/test_main.sh]: Done

File: conf/config.yaml (hunk 6 above)
Comment (kevinw66) [reply conf/config.yaml]: Done

File: apisix/cli/file.lua (hunk 7 above)
Comment (kevinw66) [reply apisix/cli/file.lua]: Done

**Review by spacewander — APPROVED** [review]

**Comment by moonming** [comment]
@kevinw66 great work 👍

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**